src/style-spec/function/expression.js

```diff
@@ -0,0 +1,362 @@
+        const parsedArgs: Array<Expression> = [];
+        for (const arg of args) {
+            parsedArgs.push(parseExpression(arg, context.concat(1 + parsedArgs.length, op)));
+        }
+
+        return new this(context.key, this.type(), parsedArgs);
+    }
+}
+
+class Reference extends BaseExpression {
+    name: string;
+    constructor(key: string, name: string, type: Type) {
+        super(key, type);
+        if (!/^[a-zA-Z_]+[a-zA-Z_0-9]*$/.test(name))
+            throw new ParsingError(key, `Invalid identifier ${name}.`);
+        this.name = name;
+    }
+
+    compile() { return this.name; }
+
+    serialize() {
+        return [this.name];
+    }
+}
+
+class LetExpression extends BaseExpression {
+    bindings: Array<[string, Expression]>;
+    result: Expression;
+    constructor(key: string, bindings: Array<[string, Expression]>, result: Expression) {
+        super(key, result.type);
+        this.bindings = [].concat(bindings);
+        this.result = result;
+    }
+
+    compile() {
+        const names = [];
+        const values = [];
+        const errors = [];
+        for (const [name, expression] of this.bindings) {
+            names.push(name);
+            const value = expression.compile();
+            if (Array.isArray(value)) {
+                errors.push.apply(errors, value);
+            } else {
+                values.push(value);
+            }
+        }
+
+        const result = this.result.compile();
+        if (Array.isArray(result)) {
+            errors.push.apply(errors, result);
+            return errors;
+        }
+
+        if (errors.length > 0) return errors;
+
+        return `(function (${names.join(', ')}) {
+            return ${result};
+        }.bind(this))(${values.join(', ')})`;
+    }
+
+    serialize() {
+        const serialized = ['let'];
+        for (const [name, expression] of this.bindings) {
+            serialized.push(name, expression.serialize());
+        }
+        serialized.push(this.result.serialize());
+        return serialized;
+    }
+
+    visit(fn: (BaseExpression) => void): void {
+        fn(this);
+        for (const binding of this.bindings) {
+            binding[1].visit(fn);
+        }
+        this.result.visit(fn);
+    }
+
+    static parse(args: Array<mixed>, context: ParsingContext) {
+        if (args.length < 3)
+            throw new ParsingError(context.key, `Expected at least 3 arguments, but found ${args.length} instead.`);
+
+        const bindings: Array<[string, Expression]> = [];
+        for (let i = 0; i < args.length - 1; i += 2) {
+            const name = args[i];
+            const key = context.path.concat(i + 1).join('.');
+            if (typeof name !== 'string')
+                throw new ParsingError(key, `Expected string, but found ${typeof name} instead`);
+
+            if (context.definitions[name])
+                throw new ParsingError(key, `"${name}" is reserved, so it cannot not be used as a "let" binding.`);
```

Comment:
Since this is getting compiled to `function (name, ...) { ... }`, we need to either prohibit JS keywords as well, or sanitize names, say by prefixing them with `_`.